File: MarkdownIOS/SimpleMarkdownParser/Classes/SymbolFinder/MarkdownSymbol.swift
//
//  MarkdownSymbol.swift
//  SimpleMarkdownParser Pod
//
//  Library symbol parsing: used to define symbols in a markdown document, like header and text style markers
//

// The enum to define the type of supported markdown symbol markers
public enum MarkdownSymbolType {
    
    case escape
    case doubleQuote
    case textBlock
    case newline
    case header
    case firstTextStyle
    case secondTextStyle
    case thirdTextStyle
    case orderedListItem
    case unorderedListItem
    case openLink
    case closeLink
    case openUrl
    case closeUrl
    
    public func isTextStyle() -> Bool {
        return self == .firstTextStyle || self == .secondTextStyle || self == .thirdTextStyle
    }
    
}

// Symbol object
public class MarkdownSymbol {
    
    // --
    // MARK: Members
    // --

    public let type: MarkdownSymbolType
    public let line: Int
    public let linePosition: Int
    public let startPosition: Int
    public let startIndex: String.Index
    public var endPosition: Int
    public var endIndex: String.Index
    

    // --
    // MARK: Initialization
    // --

    public init(type: MarkdownSymbolType, line: Int, startPosition: Int, startIndex: String.Index, endPosition: Int, endIndex: String.Index, linePosition: Int) {
        self.type = type
        self.line = line
        self.startPosition = startPosition
        self.startIndex = startIndex
        self.endPosition = endPosition
        self.endIndex = endIndex
        self.linePosition = linePosition
    }
    

    // --
    // MARK: Update position
    // --

    public func updateEndPosition(_ position: Int, index: String.Index) {
        self.endPosition = position
        self.endIndex = index
    }
    
}

File: MarkdownAndroid/SimpleMarkdownParserLib/src/main/java/com/crescentflare/simplemarkdownparser/symbolfinder/MarkdownSymbol.java
package com.crescentflare.simplemarkdownparser.symbolfinder;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Simple markdown parser library: symbol
 * Used to define symbols in a markdown document, like header and text style markers
 */
public class MarkdownSymbol implements Comparable<MarkdownSymbol>
{
    // --
    // Type enum
    // --

    public enum Type {
        Escape,
        DoubleQuote,
        TextBlock,
        Newline,
        Header,
        FirstTextStyle,
        SecondTextStyle,
        ThirdTextStyle,
        OrderedListItem,
        UnorderedListItem,
        OpenLink,
        CloseLink,
        OpenUrl,
        CloseUrl;

        public boolean isTextStyle() {
            return this == FirstTextStyle || this == SecondTextStyle || this == ThirdTextStyle;
        }
    }


    // --
    // Members
    // --

    @NotNull public Type type;
    public int line;
    public int linePosition;
    public int startPosition;
    public int endPosition;


    // --
    // Initialization
    // --

    public MarkdownSymbol(@NotNull Type type, int line, int startPosition, int endPosition, int linePosition) {
        this.type = type;
        this.line = line;
        this.startPosition = startPosition;
        this.endPosition = endPosition;
        this.linePosition = linePosition;
    }


    // --
    // Update position
    // --

    public void updateEndPosition(int position) {
        this.endPosition = position;
    }


    // --
    // Comparable implementation
    // --

    @Override
    public int compareTo(@Nullable MarkdownSymbol other) {
        if (other != null) {
            return startPosition - other.startPosition;
        }
        return 0;
    }
}
